device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/AndroidEmulatorsNode.kt

```diff
@@ -0,0 +1,352 @@
+        if (dto?.state != DeviceState.CREATED) {
+            if (dto?.state != DeviceState.FAILED) {
+                logger.error("ref=$ref is in state ${dto?.state} after request. Marking as FAILED: $dto")
+            }
+            updateDtoForRef(
+                ref = ref,
+                state = DeviceState.FAILED,
+                emulatorSerial = "",
+                model = model,
+                last_error = lastError
+            )
+            deallocateEmulatorSlot(emulatorSerial)
+        }
+    }
+
+    private fun requestEmulator(
+        ref: String,
+        emulator: EmulatorSerial,
+        model: String
+    ) {
+        logger.info("Requesting $ticketHost promise for ref=$ref emu=$emulator concurrent=$concurrentBoots")
+        httpPost("docker/request/id=$ref/device=$emulator/image=${model}")
+            .use {
+                throwUnlessOK(it)
+                val body = it.body()?.string()
+                logger.debug("Received response $it $body")
+                if (body == null || body.contains(":FAILURE:")) {
+                    logger.error("Found $ticketHost :FAILURE: or null for $emulator '${summary(body, it)}'")
+                    return
+                }
+            }
+    }
+
+    private fun summary(body: String?, response: Response): String {
+        return  if (body != null && body.isNotEmpty()) body else response.toString()
+    }
+
+    val nullRequestBody = RequestBody.create(null, "")
+
+    private fun httpPost(endpoint: String): Response {
+        return httpClient.newCall(buildEndpoint(endpoint).post(nullRequestBody).build()).execute()
+    }
+
+    private fun httpGet(endpoint: String): Response {
+        return httpClient.newCall(buildEndpoint(endpoint).build()).execute()
+    }
+
+    private fun buildEndpoint(endpoint: String) = Request.Builder().url("http://$ticketHost/$endpoint")
+
+    private fun throwUnlessOK(response: Response) {
+        if (response.code() != HttpURLConnection.HTTP_OK) {
+            throw DeviceCreationException("NOT HTTP_OK: $response")
+        }
+    }
+
+    private fun releaseOnTicketMachine(ref: DeviceRef): Boolean {
+        httpPost("docker/release/$ref")
+                .use {
+                    throwUnlessOK(it)
+                    val body = it.body()?.string()
+                    val released = body?.contains(":SUCCESS: $ref released") ?: false
+                    if (!released) {
+                        logger.warn("Unable to release $ref: $body $it")
+                    }
+                    return released
+                }
+    }
+
+    private fun emulatorOrNullFromRef(ref: DeviceRef): EmulatorSerial? {
+        val deviceDTO = refToDto[ref]
+        if (deviceDTO == null) {
+            logger.warn("No emulator for $ref")
+            return null
+        }
+        return deviceDTO.info.udid
+    }
+
+    private fun updateDtoForRef(ref: String, state: DeviceState, emulatorSerial: EmulatorSerial,
```

Comment:
Could you please put each argument on a new line

Reply:
ok